sequencing/transaction_queue.go

```diff
@@ -0,0 +1,181 @@
+package sequencing
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/dgraph-io/badger/v4"
+
+	"github.com/rollkit/go-sequencing"
+)
+
+// BatchExtender is an interface for extending a batch of transactions
+type BatchExtender interface {
+	Head(max uint64) ([]byte, error)
+	Tail(max uint64) ([]byte, error)
+}
+
+// TransactionQueue is a queue of transactions
+type TransactionQueue struct {
+	queue    []sequencing.Tx
+	mu       sync.Mutex
+	extender BatchExtender
+}
+
+// NewTransactionQueue creates a new TransactionQueue
+func NewTransactionQueue(extender BatchExtender) *TransactionQueue {
+	return &TransactionQueue{
+		queue:    make([]sequencing.Tx, 0),
+		extender: extender,
+	}
+}
+
+// GetTransactionHash to get hash from transaction bytes using SHA-256
+func GetTransactionHash(txBytes []byte) string {
+	hashBytes := sha256.Sum256(txBytes)
+	return hex.EncodeToString(hashBytes[:])
+}
+
+// AddTransaction adds a new transaction to the queue
+func (tq *TransactionQueue) AddTransaction(tx sequencing.Tx, db *badger.DB) error {
```

Comment:
similar comments as before, I think the database option should occur first before updating in memory.